scripts/content_type_stats.py

```diff
@@ -0,0 +1,109 @@
+                             'a .tsv.gz file')
+    parser.add_argument('--processes', '-P', type=int, default=1,
+                        help='number of worker processes to use (max is the '
+                             'num of cores, default: 1).')
+    parser.add_argument('--log-level', '-L', type=str, default='info',
+                        choices=['debug', 'info', 'warning',
+                                 'error', 'critical'],
+                        help='the logging level.')
+    args = parser.parse_args()
+    return args
+
+
+def process_file(input_file: Path) -> str:
+    """
+    Processes a file, collecting mime-type data from the headers of the docs
+    contained in that file.
+    It returns a string which is a tsv: each document is a single line and the
+    values are separated by tabs within the line.
+    """
+    results = ""
+    matcher_ct = re.compile(r'Content-Type: ([\w/]+)', re.I)
+    matcher_wi = re.compile(r'WARC-IDentified-Payload-Type: ([\w/]+)', re.I)
+    matcher_cd = re.compile(r'(Content-Disposition: )([^\n^"]*)"([^\n^"]*)"', re.I)
+    for doc in parse_file(input_file):
+        # Get the type from the metadata of the doc tag:
+        if doc.attrs['mime-type']:
+            type_from_doctag = doc.attrs['mime-type']
+        else:
+            type_from_doctag = '-'
+        # Get the type from the warc request header:
+        match_warcid = matcher_wi.search(doc.meta['request'])
+        if match_warcid:
+            type_from_warc_id = match_warcid.groups(1)[0]
+        else:
+            type_from_warc_id = '-'
+        # Get the type from the response header:
+        match_resp = matcher_ct.search(doc.meta['response'])
+        if match_resp:
+            type_from_response = match_resp.groups(1)[0]
+        else:
+            type_from_response = '-'
+        # Get the info about the attached file, if any:
+        match_cd = matcher_cd.search(doc.meta['response'])
+        if match_cd:
+            attached_file = match_cd.group(3)
+            if '.' in attached_file:
+                attachment_ext = attached_file.split('.')[-1]
+            else:
+                attachment_ext = 'no_extension'
+        else:
+            attachment_ext = '-'
+        # Process the results for this doc:
+        line = '\t'.join((str(input_file), type_from_doctag,
+                          type_from_warc_id, type_from_response,
+                          attachment_ext,))
+        results += line + '\n'
+    return results
+
+
+def main():
+    args = parse_arguments()
+
+    logging.basicConfig(
+        level=getattr(logging, args.log_level.upper()),
+        format='%(asctime)s - %(process)s - %(levelname)s - %(message)s'
+    )
+
+    args.output_file.parent.mkdir(parents=True, exist_ok=True)
+    with openall(args.output_file, "wt", encoding="utf-8") as f:
+        with Pool(args.processes) as p:
+            for stats in p.imap_unordered(process_file, args.input_dir.iterdir()):

```

Comment:
Not for now, but we'll have to write a function sometime that filters hidden / temporary / etc. files from directory listings so that e.g. editor swap files don't crash the scripts.

Reply:
Ok.